# <center> Python for Data Analysis</center>

In [1]:
%%HTML
<style>
td,th {
  font-size: 20px
}
</style>


## <font color=green>Table of Contents</font>
- [Quick Introduction to Pandas](#Intro)
- [DataFrame and Series](#dfs)  
    - [Series](#Series)  
    - [DataFrame](#DataFrame)
- [Loading Files](#loading)  
    - [The NYC flights Dataset](#nyc)
- [Getting a quick look at your data](#stat)  
- Filtering a DataFrame
- Sorting
- Selecting Multiple columns
- Renaming columns
- Rearranging columns
- Creating new columns
- Grouping in Pandas
- The Axis parameter
- The Apply Function
- Working with Null Values
- Reshaping/Pivoting

# <a id="Intro"> Quick Introduction to PANDAS</a>

 <b>There are many packages for data analysis in Python.  </b>
 
 To give some examples:
 
    - Different data analysis packages (Pandas and the up and coming package called Polars)
    - Linear Algebra package: Numpy, and you could include tensorflow and pytorch for deep learning.
    - Different Visualization packages (Matplotlib, Seaborn, Plotly, Altair, etc).
    
In this class, we will focus on Pandas and (depending on time) Altair and Seaborn.


## What is Pandas?

Pandas was designed to make data analysis practical and easy in Python.  Pandas can be used for data such as:  
    - Tabular Data  
    - Time-series data  
    - Numerical matrices.  

In this class, we will be focusing on <b>Pandas version 2</b>, which was released only last year.  The main difference is speed, as it allows for pyarrow support. By default, Numpy is the default backend.

Before we can analyze data using Pandas, we need to first import the package. Here is how we do it.  Note the version number. Let's also import numpy and Altair right now.

In [2]:
import pandas as pd
import numpy as np
import re
import altair as alt
import seaborn as sns
import matplotlib.pyplot as plt
import seaborn.objects as so
alt.data_transformers.disable_max_rows()
import matplotlib.pyplot as plt
print(f"Pandas:{pd.__version__}")
print(f"Altair:{alt.__version__}")
print(f"numpy:{np.__version__}")
print(f"seaborn:{sns.__version__}")

Pandas:2.1.4
Altair:5.2.0
numpy:1.23.5
seaborn:0.12.2


Here are the library versions we are using (which you can see in the output above)
 - Pandas:2.1.4
 - Altair: 5.2.0
 - Numpy: 1.23.5
 - Seaborn: 0.12.2

I will show you how to load a csv file using Pandas, but before I show you how to do that, let's discuss the data structures available in Pandas.



# <a id="dfs"> DataFrame and Series </a>

There are 2 types main data structures in Pandas: DataFrame and Series.  Let's start with the simplest data structure: Series.  
## <a id="Series"> Pandas Series</a>
The simplest data structure is called <b>Series</b>.  Think of Series as <b>vector in base R</b>.  How do we create this data structure?

We can create a Series data structure using a list, or a tuple.  We can also provide index names (optional) as follows:

In [3]:
test=pd.Series([1,2,3,4],index=["a","b","c","d"])

In [4]:
type(test)

pandas.core.series.Series

In [5]:
test

a    1
b    2
c    3
d    4
dtype: int64

Not providing index names will <b>automatically</b> create one, starting from zero, as the following example indicates:

In [6]:
pd.Series([1,2,3,4])

0    1
1    2
2    3
3    4
dtype: int64

The <b> default </b>datatype is <b> int64</b> in the example above, but we can also pass the datatype keyword.  As an example below, we are using int32.

In [7]:
pd.Series([1,2,3,4],index=["a","b","c","d"],dtype="int32")

a    1
b    2
c    3
d    4
dtype: int32

Here, instead of int64, the data type is int 32. A 64-bit integer can hold a larger number than a 32-bit integer.  So depending on your use-case, you may want to control the data type used.

The different data types available in Pandas are:

###    1. object
###    2. int (int8 to int 64, as well as unsigned integers such as  uint8 to uint64. Unsigned integers are non-negative)
###    3. float
###    4. bool
###    5. timedelta
###    6. category 

As I explained above, Series can be instantiated as a <b> Tuple </b>.

In [8]:
pd.Series((1,2,3,4))

0    1
1    2
2    3
3    4
dtype: int64

A Series can also be instantiated using a dictionary or numpy.  

Here, I have generated a Series by using <b> Dictionary </b>. The Dictionary <b> Key </b> is the index, while the <b> Value </b> is the actual value.

In [9]:
pd.Series({"a":1,"b":2,"c":3,"d":4})

a    1
b    2
c    3
d    4
dtype: int64

Here, I am instantiating a Series by using a <b> numpy </b> array.

In [10]:
pd.Series(np.array([1,2,3,4]),index=["a","b","c","d"])

a    1
b    2
c    3
d    4
dtype: int32

Now let's go back to our <b>test</b> example.  To remind you, we stored a Series in the variable test.

In [11]:
test=pd.Series([1,2,3,4],index=["a","b","c","d"])

How can we convert Pandas <b> Series </b> to <b> Numpy </b>.  The method we use is called <b> to_numpy() </b>.

In [12]:
test.to_numpy()

array([1, 2, 3, 4], dtype=int64)

Similarly, <b> to_list </b> can be used to convert a Pandas <b> Series</b> to a <b>list</b>.

In [13]:
test.to_list()

[1, 2, 3, 4]

The <b> list()</b> function works just fine as well.

In [14]:
list(test)

[1, 2, 3, 4]

And as you can imagine already, the Pandas <b>Series</b> can be converted to a <b> Dictionary </b>

In [15]:
test.to_dict()

{'a': 1, 'b': 2, 'c': 3, 'd': 4}

A <b> to_tuple() method does not exist</b>, but you can first convert to a list, and then to tuple as follows.

In [16]:
tuple(test.to_list())

(1, 2, 3, 4)

You can access values in your Series by using the index name.

In [17]:
test['a']

1

In [18]:
test['c']

3

## <a id="DataFrame">  Pandas DataFrame  </a>
A Pandas DataFrame is similar to <b> data frames </b> or <b> tibbles</b> in R.  It is used for working with data in tabular format.  A dataframe can be created from a numpy array, lists, dictionaries, Pandas series, etc as the examples below show.

In this example, I am using a DataFrame.  The column names are the <b> keys</b>, and the <b> values</b> are passed as lists.

First, let's create a Dictionary.

In [19]:
a={"one":pd.Series([1,2,3,4,5],['a','b','c','d','e']),
   "two":pd.Series([9,8,7,6,5],['a','b','c','d','e'])}

Next, we create a Pandas <b> DataFrame </b> by using the DataFrame() method.

In [20]:
a_df=pd.DataFrame(a)

The <b> head() </b> method can be used for looking at the top 5 rows.  Note that the header is also printed.

In [21]:
a_df.head()

one  two
a    1    9
b    2    8
c    3    7
d    4    6
e    5    5

Similarly, the <b> tail() </b> method can be used for looking at the bottom 5 rows (by default)

In [22]:
a_df.tail()

one  two
a    1    9
b    2    8
c    3    7
d    4    6
e    5    5

Note that by passing a number within head or tail, you can control the number of rows to show.  In the example below, we show only 2 rows.

In [23]:
a_df.head(2)

one  two
a    1    9
b    2    8

In the example below, instead of a Pandas Series, we are passing the <b> List </b> data structure.

In [24]:
a={"one":[1,2,3,4,5],
   "two":[9,8,7,6,5]}

We can then specify index labels when creating a <b> Pandas DataFrame</b> by using the index keyword.

The <b> Index labels </b> are likes rownames in R's dataframe.

In [25]:
pd.DataFrame(a,index=['a','b','c','d','e'])

one  two
a    1    9
b    2    8
c    3    7
d    4    6
e    5    5

In this example, I am passing a <b> list </b> of <b> Dictionaries </b>.  And again, the <b> keys</b> are the column names.

In [26]:
a=[{"one":1,"two":9},{"one":2,"two":8},{"one":3,"two":7},{"one":4,"two":6},{"one":5,"two":5}]

In [27]:
pd.DataFrame(a,index=['a','b','c','d','e'])

one  two
a    1    9
b    2    8
c    3    7
d    4    6
e    5    5

You can also convert a Series to a Pandas DataFrame <b> Directly </b>.  Here, we are not using a dictionary of any kind.

In [28]:
a=pd.Series([1,2,3,4,5],name="one")

In [29]:
a

0    1
1    2
2    3
3    4
4    5
Name: one, dtype: int64

The Series above has a name associated with it (one).  When you convert the Series to a DataFrame, the <b> column name<b> of the DataFrame  is the <b>Series name</b>.

In [30]:
pd.DataFrame(a)

one
0    1
1    2
2    3
3    4
4    5

# <a id="loading"> Loading/Writing files in pandas </a>

There are many different filetypes that Pandas can read.  
As an example, the <b> read_csv() </b> function can be used to quickly load csv files. Other filetypes can be loaded as well, but here we will be working on csv. We are going to use a dataset called "mpg", which is actually part of the ggplot2 library.

Anyway, here are some of the different filetypes that are supported in Pandas.

|  Type | Description | Reader | Writer |
| :---: | :---: | :---: | :---: |
|Text|CSV|read_csv|to_csv|
|Text|fixed-width text file|read_fwf|NA|
|text|JSON|read_json|to_json|
|text|HTML|read_html|to_html|
|text|Latex|NA|Styler.to_latex|
|text|XML|read_xml|to_xml|
|text|Local ClipBoard|read_clipboard|to_clipboard|
|binary|MS Excel|read_excel|to_excel|
|binary|OpenDocument|read_excel|NA|
|binary|HDF5 Format|read_HDF|to_hdf|
|binary|Feather Format|read_feather|to_feather|
|binary|Parquet Format|read_parqet|to_parquet|
|binary|ORC Format|read_orc|to_orc|
|binary|Stata|read_stata|to_stata|
|binary|SAS|read_sas|NA|
|binary|SPSS|read_spss|NA|
|binary|Python Pickle Format|read_pickle|to_pickle|
|SQL|SQL|read_sql|to_sql|
|SQL|Google BigQuery|read gbq|to_gbq|

## <a id="nyc"> Loading The NYC Flights 2013 dataset</a>

In [31]:
#flights=pd.read_csv("flights.csv")
#flights.to_csv("flights.csv.gz",compression="gzip")

The "flights" dataset is part of the tidyverse package. It can be found below.  
https://github.com/tidyverse/nycflights13?tab=readme-ov-file

The "flights" dataset has been uploaded to <b> Github </b>. As you can see from the commented lines above, the original data is uncompressed.  
I then used the <b>to_csv </b> method to save it as compressed.  
Anyway, let's load it. 

In [32]:
flights=pd.read_csv("https://github.com/niradsp/Python-for-Data-Analysis/raw/main/flights.csv.gz",compression="gzip",index_col=0)

This is a <b> compressed csv </b> (comma separated) file.  
The first thing you notice is that Pandas allows you to <b> read directly from the web.</b> 
In the above example, I read from my github page.  The flights dataset is a big file (~25 MB in total) when unzipped.
So I saved it as compressed.  You can read the gzipped file if you use the compression="gzip" parameter.

Some other parameters that can be changed are:  
- <b> sep</b>:  The separator used.
- <b> header </b>:  Row Number of Header.
- <b> skiprow </b>:  Number of lines to skip
- <b> nrows </b>: Number of lines to read.

<b> And many more </b>

You can just go to the documentation below to know about the different parameters you can control.  
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html


We will shart by looking at the dimension of the flights object, along with a quick glance at what the data looks like.

This is handled by the <b> shape </b> attribute.

In [33]:
flights.shape

(336776, 19)

There are 336,776 rows of data, along with 19 columns.

In [34]:
flights.head(3)

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0  2013      1    1     517.0             515        2.0     830.0   
1  2013      1    1     533.0             529        4.0     850.0   
2  2013      1    1     542.0             540        2.0     923.0   

   sched_arr_time  arr_delay carrier  flight tailnum origin dest  air_time  \
0             819       11.0      UA    1545  N14228    EWR  IAH     227.0   
1             830       20.0      UA    1714  N24211    LGA  IAH     227.0   
2             850       33.0      AA    1141  N619AA    JFK  MIA     160.0   

   distance  hour  minute      time_hour  
0      1400     5      15  1/1/2013 5:00  
1      1416     5      29  1/1/2013 5:00  
2      1089     5      40  1/1/2013 5:00

The  <b> columns</b>  are: 
- year: Year of flight.  In this case, it is 2013.
- month: Month of flight
- Day:  Departure day
- Dep_time,Arr_time:  Departure Time and Arrival Time (format=HHMM or HMM) in local timezone.
- Sched_arr_time and Sched_dep_time:  Sheduled arrival and departure times.
- Carrier:  2-letter Carrier abbreviation (for example, UA=United Airlines)  
- flight:  Flight number
- tailnum:  Plane tailnumber.
- orig, des:  Origin and Destination. For example, JFK would be John F Kennedy in New York.
- air_time:  Time spent in the air, in minutes.
- distance:  Distance between airports, given in miles.
- hour, minute: Departure time broken into hour and minute.
- time_hour:  Scheduled Date and Hour of the flight.

# <a id="stat">Getting a quick look at your data</a>

You can use the <b> info </b> metho to see if there are any null values in your dataframe.  This command is probably one of the first commands you should use when performing exploratory data analysis (EDA).There is also the Dtype column.

In [35]:
flights.info()

<class 'pandas.core.frame.DataFrame'>
Index: 336776 entries, 0 to 336775
Data columns (total 19 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   year            336776 non-null  int64  
 1   month           336776 non-null  int64  
 2   day             336776 non-null  int64  
 3   dep_time        328521 non-null  float64
 4   sched_dep_time  336776 non-null  int64  
 5   dep_delay       328521 non-null  float64
 6   arr_time        328063 non-null  float64
 7   sched_arr_time  336776 non-null  int64  
 8   arr_delay       327346 non-null  float64
 9   carrier         336776 non-null  object 
 10  flight          336776 non-null  int64  
 11  tailnum         334264 non-null  object 
 12  origin          336776 non-null  object 
 13  dest            336776 non-null  object 
 14  air_time        327346 non-null  float64
 15  distance        336776 non-null  int64  
 16  hour            336776 non-null  int64  
 17  minute         

You can also use the <b> describe()</b> method to quickly look at some summary statistics.

In [36]:
flights.describe()

year          month            day       dep_time  sched_dep_time  \
count  336776.0  336776.000000  336776.000000  328521.000000   336776.000000   
mean     2013.0       6.548510      15.710787    1349.109947     1344.254840   
std         0.0       3.414457       8.768607     488.281791      467.335756   
min      2013.0       1.000000       1.000000       1.000000      106.000000   
25%      2013.0       4.000000       8.000000     907.000000      906.000000   
50%      2013.0       7.000000      16.000000    1401.000000     1359.000000   
75%      2013.0      10.000000      23.000000    1744.000000     1729.000000   
max      2013.0      12.000000      31.000000    2400.000000     2359.000000   

           dep_delay       arr_time  sched_arr_time      arr_delay  \
count  328521.000000  328063.000000   336776.000000  327346.000000   
mean       12.639070    1502.054999     1536.380220       6.895377   
std        40.210061     533.264132      497.457142      44.633292   
min       -43.000000       1.000000        1.000000     -86.000000   
25%        -5.000000    1104.000000     1124.000000     -17.000000   
50%        -2.000000    1535.000000     1556.000000      -5.000000   
75%        11.000000    1940.000000     1945.000000      14.000000   
max      1301.000000    2400.000000     2359.000000    1272.000000   

              flight       air_time       distance           hour  \
count  336776.000000  327346.000000  336776.000000  336776.000000   
mean     1971.923620     150.686460    1039.912604      13.180247   
std      1632.471938      93.688305     733.233033       4.661316   
min         1.000000      20.000000      17.000000       1.000000   
25%       553.000000      82.000000     502.000000       9.000000   
50%      1496.000000     129.000000     872.000000      13.000000   
75%      3465.000000     192.000000    1389.000000      17.000000   
max      8500.000000     695.000000    4983.000000      23.000000   

              minute  
count  336776.000000  
mean       26.230100  
std        19.300846  
min         0.000000  
25%         8.000000  
50%        29.000000  
75%        44.000000  
max        59.000000